### ISSUE TITLE
ngMockE2E is not available

### ISSUE THREAD

**Issue by ghost** [yevhenpavliuk/ng-mock-e2e#1]
Hello,
first of all thanks a lot for this library, it's working great. However I noticed a random issue when I run E2E test. Sometimes the module is not available and I guess it's because the embedScript is running in parallel of the addMockModule, and probably, sometimes, the script is not injected into the browser when try to add the module. I think you should wait until the bower dependency has been loaded into the page first and the start using the module.

```
beforeEach(function () {
  ngMockE2E.addMockModule();
  ngMockE2E.addAsDependencyForModule('myApp');
  ngMockE2E.embedScript('/bower_components/angular-mocks/angular-mocks.js'); // This should be the cause of the problem
});
```

This is my random error:
- Failed: Error while running module script ngMockE2E_: unknown error: [$injector:nomod] Module 'ngMockE2E' is not available! You either misspelled the module name or forgot to load it. If registering a module ensure that you specify the dependencies as the second argument.
  http://errors.angularjs.org/1.4.3/$injector/nomod?p0=ngMockE2E
  (Session info: chrome=49.0.2623.87)
  (Driver info: chromedriver=2.16.333243 (0bfa1d3575fc1044244f21ddb82bf870944ef961),platform=Mac OS X 10.11.3 x86_64) (WARNING: The server did not provide any stacktrace information)
  Command duration or timeout: 4 milliseconds
  Build info: version: '2.46.0', revision: '87c69e2', time: '2015-06-04 16:16:47'
  System info: host: 'rnln-iosbuild.resilientplc.priv', ip: '172.16.3.74', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.11.3', java.version: '1.7.0_51'
  Session ID: 300b93297be4dee2d6879e4228bad876
  Driver info: org.openqa.selenium.chrome.ChromeDriver
  Capabilities [{platform=MAC, acceptSslCerts=true, javascriptEnabled=true, browserName=chrome, chrome={userDataDir=/var/folders/dx/1b5b1yq11gn8bs_qyt42gj3h00007h/T/.org.chromium.Chromium.dfxVEV}, rotatable=false, locationContextEnabled=true, mobileEmulationEnabled=false, version=49.0.2623.87, takesHeapSnapshot=true, cssSelectorsEnabled=true, databaseEnabled=false, handlesAlerts=true, browserConnectionEnabled=false, webStorageEnabled=true, nativeEvents=true, hasTouchScreen=false, applicationCacheEnabled=false, takesScreenshot=true}]

**Comment by yevhenpavliuk** [comment]
@azornada-resilient 
Hello Andrea,
Sorry I didn't reply earlier - I just saw the issue and there was no email sent to me regarding that.
I'll take a look at it this weekend.
Thanks for your feedback!

**Comment by yevhenpavliuk** [comment]
Didn't get to it this weekend. I'll try again in a week.

**Comment by andreasonny83** [comment]
no worry,
just let me know if you need more information when you're ready

**Comment by yevhenpavliuk** [comment]
@andreasonny83

> script is not injected into the browser when try to add the module

That's exactly the reason.

I felt wrong when I added `embedScript` to the library. This method is out of scope of the library's responsibilities.

I think it should be deprecated. Instead, adding `angular-mocks.js` should be handled outside the library. For instance, it can be added by client code based on a query string parameter:

``` html
<script>
if (/(?:\?|&)protractor(?:&|$)/.test(location.search)) {
  document.write('<script src="https://code.angularjs.org/1.5.5/angular-mocks.js"></' + 'script>');
}
</script>
```

Or maybe there's a library that allows including scripts so that Protractor waits until they load. In this case just use such a library.

**Comment by andreasonny83** [comment]
I agree it shouldn't be a library responsibility and it shouldn't force to include the angular mock because your library doesn't know which version of Angular and angular-mock the user want to use on his/her project.
Instead, what you can do in order to optimise your library, could be to verify if the angular-mock has been included in the project and, if not, disable the library and report a warning to the user.

**Comment by yevhenpavliuk** [comment]
@andreasonny83
Sounds reasonable. Thanks!
I'll reserve some time to work on this during the week (I have vacation now).
I'll keep you updated.

**Comment by bladyk1** [comment]
Hi, I have exactly same error mentioned in 1st post, I'm testing UI and I would like to mock https request the way is described in second post here https://github.com/angular/protractor/issues/125
I have the mocked-backend.js including code:

```javascript
var ngMockE2E = require('ng-mock-e2e');

exports.httpBackendMock = function() {
    angular.module('httpBackendMock', ['app', 'ngMockE2E'])
        .run(function ($httpBackend) {
            $httpBackend.whenGET('..config.json')
                .respond([
                    {
                        ...
                    }
                ]);
        });
};
```
then in onPrepare hook I have:
```javascript
    var mockModule = require('./mocked-backend');
    browser.addMockModule('httpBackendMock', mockModule.httpBackendMock);
```

and angular-mocks is added as a dependency to package.json
```javascript
    "angular-mocks": "1.6.5"
```

Can anyone help me with this please? Do I have to modify website code to make it work ?

**Comment by yevhenpavliuk** [comment]
Hi @bladyk1,

Let me look into the issue.

Sorry, I've been neglecting it for a long time, but I hope I've gained some enthusiasm to finally start working on it.

I'll try to do what I can and write you back here soon.

**Comment by bladyk1** [comment]
ok, thanks

**Comment by yevhenpavliuk** [comment]
Hey @bladyk1,

You don't need `ngMockE2E` node module in your case. I think you just need to embed `angular-mocks.js` and add it as a dependency for your app. Try this:

```javascript
exports.httpBackendMock = function() {
    angular.module('app').requires.push('ngMockE2E'); // Add ngMockE2E from angular-mocks.js as a dependency for your app.
    angular.module('app').run(function ($httpBackend) {
            $httpBackend.whenGET('..config.json')
                .respond([
                    {
                        ...
                    }
                ]);
        });

    // Embed angular-mocks.js
    const script = document.createElement('script');
    script.src = 'bower_components/angular-mocks/angular-mocks.js'; // The actual needs to be adjusted for your project.
    document.body.appendChild(script);
};
```

```javascript
    var mockModule = require('./mocked-backend');
    browser.addMockModule('app', mockModule.httpBackendMock);
```

And don't forget to remove the module in `afterEach`:

```javascript
    afterEach(() => {
        browser.removeMockModule('app');
    });
```

As you can see, you can avoid creating an extra module `httpBackendMock`.

Nevertheless, please, check out the updated [README](https://github.com/yevhenpavliuk/ng-mock-e2e/blob/master/README.md) and [this tiny example project](https://github.com/yevhenpavliuk/ng-mock-e2e-example) I made this weekend. I hope it'll give you better insights. Please, tell me if not, and I'll try to help you further.

**Comment by yevhenpavliuk** [comment]
Hello @azornada-resilient (@andreasonny83),

Sorry for this tardy reply. I lost enthusiasm back then because my daily work changed so that there was no need in using this module, and I thought that other people didn't really use it. But recent issues and download numbers gave me a hope that it is not completely dead. So I got some enthusiasm back to continue working on the project.

I start looking into the issue, but I couldn't reproduce it. I made [this tiny example project](https://github.com/yevhenpavliuk/ng-mock-e2e-example) during this weekend.

I found out that if you remove a mock module at the end of `it`, it prevents all functions added under the name from running:

```javascript
escribe('example without ngMockE2E', () => {
  beforeEach(() => {
    browser.addMockModule('example', () => {
      angular.module('example').requires.push('ngMockE2E');

      const script = document.createElement('script');
      script.src = 'bower_components/angular-mocks/angular-mocks.js';
      document.body.appendChild(script);
    });
  });

  afterEach(() => {
    browser.removeMockModule('example'); // For the mock module that's added in `beforeEach`.
    browser.removeMockModule('example'); // For the mock module that's added in `it`.
    // The number of removals should be equal to the number of additions.
    // Removing the mock module in `it` prevents all registered functions from execution.
  });

  it('should have heading "It works!" if the server responds "It works!"', () => {
    browser.addMockModule('example', () => {
      angular.module('example').
        run($httpBackend => {
          $httpBackend.when('GET', 'heading').respond('It works!');
        });
    });

    browser.get('/');
    expect($('h1').getText()).toEqual('It works!');

    // browser.removeMockModule('example'); // Removing the mock module in `it` prevents all registered functions from execution.
  });
});
```

It's not the issue you described, but it looked somewhat related, so I decided to let you know. In case your problem is still actual, please, give me an example of your project organization where I could reproduce the issue.

**Comment by bladyk1** [comment]
Hi @yevhenpavliuk
Really  appreciate your help. I see most of examples I found till now about mocking http request is used with Jasmine. I'm using Protractor with Cucumber, but it should work I think

I changed my code to yours accordingly
Instead of
```javascript
script.src = 'bower_components/angular-mocks/angular-mocks.js';
```
I have
```javascript
script.src = '/node_modules/angular-mocks/angular-mocks.js';
```
and now I received error:
Error: Error while running module script app: unknown error: 'script' must be a string
What else can be wrong ?

**Comment by yevhenpavliuk** [comment]
@azornada-resilient (@andreasonny83),

I've found how to reproduce the issue. You were right. It happens if downloading of `angular-mocks.js` takes longer. For instance, if the file is taken from a remote CDN instead of the local server.

I haven't come up with a solution yet. The idea is to have `angular-mocks.js` ready before the app bootstraps. So, yeah, as it's been mentioned before, `embedScript` isn't good enough.

**Comment by yevhenpavliuk** [comment]
@bladyk,

I have an assumption. The example code uses ES6 features (`const` in particular). Maybe your browser environment doesn't support that. Try replacing `const` with `var`.

Other than that, you're using `/node_modules/angular-mocks/angular-mocks.js` as the URL. Is the file available on your host?

Alternatively, just to make sure it works for you, try to put the entire contents of `angular-mocks.js` as a mock module:

```javascript
browser.addMockModule('example', function() {
  // Copy and paste the source code of https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.5.0/angular-mocks.js here.
});
```

Instead of embedding a local script:

```javascript
browser.addMockModule('example', function() {
  const script = document.createElement('script');
  script.src = '/node_modules/angular-mocks/angular-mocks.js';
  document.body.appendChild(script);
});
```
